synapse/storage/databases/main/events_worker.py

```diff
@@ -657,10 +662,18 @@ async def get_missing_events_from_db() -> Dict[str, EventCacheEntry]:
                 # the events have been redacted, and if so pulling the redaction event
                 # out of the database to check it.
                 #
+                missing_events = {}
                 try:
-                    missing_events = await self._get_events_from_db(
+                    # First fetch from the cache - including any external caches
+                    cache_missing_events = await self._get_events_from_cache(
                         missing_events_ids,
                     )
+                    missing_events.update(cache_missing_events)
+                    # Now actually fetch any remaining events from the DB
+                    db_missing_events = await self._get_events_from_db(
+                        missing_events_ids - set(cache_missing_events.keys()),
+                    )
```

Comment:
I think there's some redundant dict copying going on here:

```suggestion
                    missing_events = await self._get_events_from_cache(
                        missing_events_ids,
                    )
                    # Now actually fetch any remaining events from the DB
                    db_missing_events = await self._get_events_from_db(
                        missing_events_ids - missing_events.keys(),
                    )
```